modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,628 @@
+pub type BaseCoin = Coin<Denom>;
+
+/// Base denomination type
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Display)]
+#[serde(transparent)]
+pub struct Denom(String);
+
+impl FromStr for Denom {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.trim().is_empty() {
+            Err(Error::empty_base_denom())
+        } else {
+            Ok(Denom(s.to_owned()))
+        }
+    }
+}
+
+#[derive(Clone, Debug, Ord, PartialOrd, Eq, PartialEq)]
+pub struct TracePrefix {
+    port_id: PortId,
+    channel_id: ChannelId,
+}
+
+impl TracePrefix {
+    pub fn new(port_id: PortId, channel_id: ChannelId) -> Self {
+        Self {
+            port_id,
+            channel_id,
+        }
+    }
+}
+
+impl fmt::Display for TracePrefix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}/{}", self.port_id, self.channel_id)
+    }
+}
+
+/// A full trace path modelled as a collection of `TracePrefix`s.
+#[derive(Clone, Debug, Default, Eq, PartialEq, PartialOrd, Ord, From)]
+pub struct TracePath(Vec<TracePrefix>);
+
+impl<'a> TryFrom<Vec<&'a str>> for TracePath {
+    type Error = Error;
+
+    fn try_from(v: Vec<&'a str>) -> Result<Self, Self::Error> {
+        if v.len() % 2 != 0 {
+            return Err(Error::invalid_trace_length(v.len()));
+        }
+
+        let mut trace = vec![];
+        let id_pairs = v.windows(2).map(|paths| (paths[0], paths[1]));
+        for (pos, (port_id, channel_id)) in id_pairs.step_by(2).enumerate() {
+            let port_id =
+                PortId::from_str(port_id).map_err(|e| Error::invalid_trace_port_id(pos, e))?;
+            let channel_id = ChannelId::from_str(channel_id)
+                .map_err(|e| Error::invalid_trace_channel_id(pos, e))?;
+            trace.push(TracePrefix {
+                port_id,
+                channel_id,
+            });
+        }
+
+        Ok(trace.into())
+    }
+}
+
+impl FromStr for TracePath {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = {
+            let parts: Vec<&str> = s.split('/').collect();
+            if parts.len() == 1 && parts[0].trim().is_empty() {
+                vec![]
+            } else {
+                parts
+            }
+        };
+        parts.try_into()
+    }
+}
+
+impl fmt::Display for TracePath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let path = self
+            .0
+            .iter()
+            .map(|prefix| prefix.to_string())
+            .collect::<Vec<String>>()
+            .join("/");
+        write!(f, "{}", path)
+    }
+}
+
+/// Indicates whether the sender chain is acting as a source or sink. Each send to any chain other
+/// than the one it was previously received from is a movement forwards in the token's timeline. In
+/// these instances the sender chain is acting as the source zone.
+pub enum Source {
```

Comment:
Would it be better to rename as
+ `Source` -> `TokenSource`
+ `Sender` -> `SenderChain`
+ `Receiver` -> `ReceiverChain`

?

Reply:
I have trouble understanding this enum, so a renaming here would be warranted. Additionally, "source" is very overloaded term.